### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by bentheiii**

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1 +1,7 @@
-<$license$>
\ No newline at end of file
+Copyright (c) 2021, ben avrahami <ben.avrahami@biocatch.com>
```
Comment (aviramha) [inline LICENSE:1]: Biocatch *

Comment (bentheiii) [reply LICENSE:1]: drat foiled again

File: doc/source/conf.py [hunk 2]
```diff
@@ -17,9 +17,9 @@
 
 # -- Project information -----------------------------------------------------
 
-project = '<$package$>'
-copyright = '<$year$>, <$authors$>'
-author = '<$authors$>'
+project = 'heksher'
+copyright = '2021, ben avrahami <ben.avrahami@biocatch.com>'
```
Comment (aviramha) [inline doc/source/conf.py:21]: biocatch

File: pyproject.toml [hunk 3]
```diff
@@ -1,25 +1,29 @@
...
+repository = "https://github.com/biocatchltd/heksher-py"
 packages = [
-    {include=<$package!quote$>}
+    {include="heksher"}
 ]
 
 [tool.poetry.dependencies]
-python = <$py_versions!semver!quote$>
+python = "^3.7"
+orjson = "^3.4.6"
```
Comment (aviramha) [inline pyproject.toml:15]: asd

Comment (aviramha) [reply pyproject.toml:15]: v

File: heksher/heksher_client.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+class BaseHeksherClient(ABC):
+    @abstractmethod
+    def handover_main(self, other: BaseHeksherClient):
+        """
+        attempt to hand over the "main"-ness of the client from self to other. Might fail if self feels this
+        shouldn't happen. Calling this when self is not the main client will result in a runtime error. Note that this
+        method should only handle the stopping and cleanup of self, and not start other.
+        """
+        if heksher.main_client.Main is not self:
```
Comment (aviramha) [inline heksher/heksher_client.py:19]: I think Handover should support only getting a temproary client, as it's not like you can transfer between heksher clients

File: heksher/heksher_client.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            raise RuntimeError('self must be the main client')
+
+    @abstractmethod
+    def add_undeclared(self, settings: Collection[Setting]):
```
Comment (aviramha) [inline heksher/heksher_client.py:23]: add_undeclared sounds odd, why not add_setting?

File: heksher/heksher_client.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        """
+        Register a setting to be declared.
+        Args:
+            settings: the setting that needs to be declared.
+        """
+        pass
+
+    @abstractmethod
+    def context_namespace(self, ns: Mapping[str, str]) -> Mapping[str, str]:
```
Comment (aviramha) [inline heksher/heksher_client.py:32]: ns -> Namespace, don't be cheap

File: heksher/clients/subclasses.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    async def close(self):
+        pass
+
+    async def __aenter__(self: T) -> T:
+        await self.set_as_main()
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        await self.close()
+        return None
```
Comment (aviramha) [inline heksher/clients/subclasses.py:211]: no need to return None

**Review by aviramha — CHANGES_REQUESTED** [review]

File: heksher/heksher_client.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+from typing import Mapping, Collection
+from weakref import WeakSet
+
+import heksher.main_client
+from heksher.setting import Setting
+
+
+class BaseHeksherClient(ABC):
+    @abstractmethod
+    def handover_main(self, other: BaseHeksherClient):
```
Comment (omerhazan) [inline heksher/heksher_client.py:13]: When will the client use this? in which scenario?

File: heksher/clients/subclasses.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        """
+        response.raise_for_status()
+        response_data = response.json()
+        incomplete = response_data['incomplete']
+        if incomplete:
+            logger.warning('some setting entries are incomplete in declaration',
+                           extra={'setting': setting.name, 'incomplete': incomplete})
+        changed = response_data['changed']
+        if changed:
+            logger.warning('some setting entries were changed by declaration',
```
[the remark this answers is not in the document]
Comment (bentheiii) [reply heksher/clients/subclasses.py:99]: Because it might be desired and might not

**Review by omerhazan — APPROVED** [review]

**Review by bharel — CHANGES_REQUESTED** [review]
I'm around halfway through

File: heksher/clients/async_client.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+                data['cache_time'] = self._last_cache_time.isoformat()
+            self._last_cache_time = datetime.now()
+
+            response = await self._http_client.request('GET', '/api/v1/rules/query', data=orjson.dumps(data))
+            response.raise_for_status()
+
+            updated_settings = response.json()['rules']
+            async with self.modification_lock:
+                self._update_settings_from_query(updated_settings)
+            logger.info('heksher reload done', extra={'updated_settings': list(updated_settings.keys())})
```
[the remark this answers is not in the document]
Comment (bentheiii) [reply heksher/clients/async_client.py:107]: I don't think this is correct. If a setting changed between line 100 and 105, I want to get it in the next loop. I'd rather get the same update twice than miss an update.

File: heksher/clients/async_client.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    async def set_as_main(self):
+        await super().set_as_main()
+
+        # check that we're dealing with the right context features
+        try:
+            response = await self._http_client.get('/api/v1/context_features/')
+            response.raise_for_status()
+        except HTTPError:
+            logger.exception('failure to get context_features from heksher service',
+                             extra={'service_url': self._service_url})
```
[the remark this answers is not in the document]
Comment (bentheiii) [reply heksher/clients/async_client.py:133]: We actually want to continue by design, we want the services to keep going even if the keksher server is unavailable.

[the remark this answers is not in the document]
Comment (aviramha) [reply heksher/clients/async_client.py:133]: I agree with Ben, I'd just add that's it's worth to expose this to the consumer. I suggest adding an "alive" attribute or something, that is False if last communication attempt resulted in an error.

File: pyproject.toml [hunk 12]
```diff
@@ -1,25 +1,29 @@
...
+license = "MIT"
 readme = "README.md"
-repository = <$repo_link!quote$>
+repository = "https://github.com/biocatchltd/heksher-py"
 packages = [
-    {include=<$package!quote$>}
+    {include="heksher"}
 ]
 
 [tool.poetry.dependencies]
```
Comment (aviramha) [inline pyproject.toml:13]: we probably need to be way more liberal on the dependencies versions, as we don't want to restrict users. 
I'd set it for same major as minimum

File: heksher/clients/async_client.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (bharel) [reply heksher/clients/async_client.py:107]: That's right. In the current scenario you miss an update.

File: heksher/clients/async_client.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (bharel) [reply heksher/clients/async_client.py:133]: So we want services to continue going even if we have the wrong configuration? That doesn't sound right...

[the remark this answers is not in the document]
Comment (bentheiii) [reply heksher/clients/async_client.py:133]: We do have a `ping` method for the clients, they can use that to check connectivity.
As for restoration, I think just restarting the service would be good enough for now. We can probably put it up as an issue to be handled later.

**Pull request merged**